internal/logger/logger.go

```diff
@@ -0,0 +1,294 @@
+package logger
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+	"syscall"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/bsontype"
+)
+
+const messageKey = "message"
+const jobBufferSize = 100
+const logSinkPathEnvVar = "MONGODB_LOG_PATH"
+const maxDocumentLengthEnvVar = "MONGODB_LOG_MAX_DOCUMENT_LENGTH"
+
+// DefaultMaxDocumentLength is the default maximum length of a stringified BSON document in bytes.
+const DefaultMaxDocumentLength = 1000
+
+// TruncationSuffix are trailling ellipsis "..." appended to a message to indicate to the user that truncation occurred.
+// This constant does not count toward the max document length.
+const TruncationSuffix = "..."
+
+// LogSink represents a logging implementation. It is specifically designed to be a subset of go-logr/logr's LogSink
+// interface.
+type LogSink interface {
+	Info(int, string, ...interface{})
+}
+
+type job struct {
+	level Level
+	msg   ComponentMessage
+}
+
+// Logger is the driver's logger. It is used to log messages from the driver either to OS or to a custom LogSink.
+type Logger struct {
+	ComponentLevels   map[Component]Level
+	Sink              LogSink
+	MaxDocumentLength uint
+
+	jobs chan job
+}
+
+// New will construct a new logger with the given LogSink. If the given LogSink is nil, then the logger will log using
+// the standard library.
+//
+// If the given LogSink is nil, then the logger will log using the standard library with output to os.Stderr.
+//
+// The "componentLevels" parameter is variadic with the latest value taking precedence. If no component has a LogLevel
+// set, then the constructor will attempt to source the LogLevel from the environment.
+func New(sink LogSink, maxDocumentLength uint, componentLevels map[Component]Level) *Logger {
+	return &Logger{
+		ComponentLevels: selectComponentLevels(
+			func() map[Component]Level { return componentLevels },
+			getEnvComponentLevels,
+		),
```

Comment:
The `New` function will not override any user-defined variables, it fills in data that was not user-defined with environment variables per the [specification](https://github.com/mongodb/specifications/blob/master/source/logging/logging.rst).

Reply:
My original comment was based on the understanding that the environment variable configuration would override the provided configuration options. However, it sounds like the environment variable configuration is only used when the passed-in values are nil/zero. That behavior sounds reasonable.

I agree with @benjirewis that writing the "environment variable fallback" conditional logic in `New` would be more obvious than calling a function because you don't have to know the function's selection precedence (see my misunderstanding above for example 😆 ).